README.rst
BrowserStacker
==============
Python wrapper for `BrowserStack <https://www.browserstack.com/>`_ features.

|Build Status| |codecov.io|


Installation
------------

The current stable release:

::

    pip install browserstacker

or:

::

    easy_install browserstacker

or from source:

::

    $ sudo python setup.py install

Usage
-----

To start working with BrowserStack ScreenShots API simply type:

.. code:: python

    >>> from browserstacker import ScreenShotsAPI
    >>> api = ScreenShotsAPI('user', 'key')
    >>> api.browsers()
    [
        {
          "os": "Windows",
          "os_version": "XP",
          "browser": "chrome",
          "browser_version": "21.0",
          "device": None
        },
        {
          "os": "ios",
          "os_version": "6.0",
          "browser": "Mobile Safari",
          "browser_version": None,
          "device": "iPhone 4S (6.0)"
        }
      ....
    ]

Optionally you can filter the output of ``browsers`` by ``browser``, ``browser_version``, ``device``, ``os`` or ``os_version``.

.. code:: python

    >>> api.browsers(os='Windows', browser='chrome')
    [
        {
          "os": "Windows",
          "os_version": "8.1",
          "browser": "chrome",
          "browser_version": "22.0",
          "device": None
        },
        {
          "os": "Windows",
          "os_version": "8.1",
          "browser": "chrome",
          "browser_version": "23.0",
          "device": None
        }
      ....
    ]

Command line:

.. code:: bash

    $ browserstacker browsers -os Windows -b chrome
    Available browsers:

    ------
    browser_version: 22.0
    os: Windows
    browser: chrome
    device: None
    os_version: 8.1
    ------
    ...
    Total browsers: 100


Filtration values are case insensitive and are casted to strings during filtration.
E.g. you can use integers for filtration by ``os_version`` parameter.
Also it is possible to filter by multiple values:

.. code:: python

    >>> api.browsers(os='Windows', browser=('chrome', 'firefox'))
    [
        {
          "os": "Windows",
          "os_version": "8.1",
          "browser": "chrome",
          "browser_version": "22.0",
          "device": None
        },
        {
          "os": "Windows",
          "os_version": "8.1",
          "browser": "firefox",
          "browser_version": "16.0",
          "device": None
        }
      ....
    ]

Command line:

.. code:: bash

    $ browserstacker browsers -os Windows -b chrome -b firefox
    Available browsers:

    ------
    browser_version: 22.0
    os: Windows
    browser: chrome
    device: None
    os_version: 8.1
    ------
    ...
    Total browsers: 215

To generate screenshots:

.. code:: python

    >>> browser = api.browsers()[0]
    >>> response = api.generate('http://www.google.com', browser)

Command line:

.. code:: bash

    $ browserstacker generate http://www.google.com -os Windows -b firefox -bv 37.0 -ov XP

Here you can pass single browsers or list of them.
You'll get the ``job_id`` from response. To list available screenshots for ``job_id``:

.. code:: python

    >>> api.list(response['job_id'])
    {
        "id":"13b93a14db22872fcb5fd1c86b730a51197db319",
        "state":"done",
        "callback_url": "http://staging.example.com",
        "win_res": "1024x768",
        "mac_res": "1920x1080",
        "quality": "compressed",
        "wait_time": 5,
        "screenshots": [
            {
                "os":"Windows",
                "os_version":"XP",
                "browser":"ie",
                "browser_version":"7.0",
                "id":"be9989892cbba9b9edc2c95f403050aa4996ac6a",
                "state":"done",
                "url":"www.google.com",
                "thumb_url":"https://www.browserstack.com/screenshots/13b93a14db22872fcb5fd1c86b730a51197db319/thumb_winxp_ie_7.0.jpg",
                "image_url":"https://www.browserstack.com/screenshots/13b93a14db22872fcb5fd1c86b730a51197db319/winxp_ie_7.0.png",
                "created_at":"2013-03-14 16:25:45 UTC",
            }
        ]
    }

Command line:

.. code:: bash

    $ browserstacker list 13b93a14db22872fcb5fd1c86b730a51197db319

To download screenshots:

.. code:: python

    >>> api.download(response['job_id'], 'path_to_screenshots_dir')

Command line:

.. code:: bash

    $ browserstacker download 13b93a14db22872fcb5fd1c86b730a51197db319 -d screenshots_dir

All screenshots will be saved in 'path_to_screenshots_dir'. If ``destination`` kwarg is absent, then screenshots will be
downloaded to current working directory.
Also you can use shortcut to create & download screenshots to your local machine:

.. code:: python

    >>> response = api.make('http://www.google.com', browser, destination='path_to_screenshots_dir')

Command line:

.. code:: bash

    $ browserstacker make -os Windows -b firefox -bv 37.0 -ov XP -d screenshots_dir

Command line interface
~~~~~~~~~~~~~~~~~~~~~~

Browserstacker comes with command line interface built with `Click <http://click.pocoo.org/>`_.
After installation ``browserstacker`` entry point will be available
It exposes almost the same API as ``ScreenShotsAPI`` does. Does not support Python 3.2.

To use CLI you have to pass user & key directly:

.. code:: bash

    $ browserstacker -u user -k key browsers
    Available browsers:
    ...
    Total browsers: 629

Or via environment variables:

.. code:: bash

    $ export BROWSERSTACK_USER=user
    $ export BROWSERSTACK_KEY=key
    $ browserstacker browsers

Help is also available:

.. code:: bash

    $ browserstacker --help
    Usage: browserstacker [OPTIONS] COMMAND [ARGS]...

    Options:
      -u, --user TEXT  Username on BrowserStack
      -k, --key TEXT   Access key
      -v, --verbosity  Verbosity level
      --version        Show the version and exit.
      --help           Show this message and exit.

    Commands:
      download
      generate
      browsers
      list
      make

Python support
--------------

BrowserStacker supports Python 2.6, 2.7, 3.2, 3.3, 3.4, 3.5, PyPy, PyPy3 and Jython.
**NOTE**. CLI does not support Python 3.2.


.. |Build Status| image:: https://travis-ci.org/Stranger6667/browserstacker.svg?branch=master
   :target: https://travis-ci.org/Stranger6667/browserstacker

.. |codecov.io| image:: https://codecov.io/github/Stranger6667/browserstacker/coverage.svg?branch=master
    :target: https://codecov.io/github/Stranger6667/browserstacker?branch=master

CHANGELOG.md
History
-------

0.3.1 - 19.06.2016

* Changed commands names.
* Fixed logger setup.
* Initial support for retries (#3).
* Dropped sessions support due auth complications.

0.3 - 11.02.2016
---------

* Added verbose output option (#9).
* Added filtration for `list_browsers` (#10).

0.2.1 - 25.01.2016
---------

* Added `make_screenshots` shortcut (#7).
* Added default_browser option (#8).

0.2 - 25.01.2016
---------

* Used single `requests` session for all calls (#1).
* Package is uploaded to PyPI (#2).
* Documentation added (#6).

0.1 - 22.01.2016
----------------

* Initial release.


browserstacker/screenshots.py
# coding: utf-8
import logging
import os
import sys
from time import sleep

import requests

from ._compat import urljoin


DEFAULT_LOGGING_LEVEL = logging.CRITICAL
DEFAULT_TIMEOUT = 10
DEFAULT_RETRIES = 6


class AuthError(RuntimeError):
    pass


class ParallelLimitReached(RuntimeError):
    pass


def get_logger(verbosity):
    """
    Returns simple console logger.
    """
    logger = logging.getLogger(__name__)
    logger.setLevel({
        0: DEFAULT_LOGGING_LEVEL,
        1: logging.INFO,
        2: logging.DEBUG
    }.get(min(2, verbosity), DEFAULT_LOGGING_LEVEL))
    logger.handlers = []
    handler = logging.StreamHandler(sys.stdout)
    handler.setLevel(logging.DEBUG)
    handler.setFormatter(logging.Formatter('%(name)s - %(levelname)s - %(message)s'))
    logger.addHandler(handler)
    return logger


def match_item(key, value, item):
    """
    Check if some item matches criteria.
    """
    if isinstance(value, (list, tuple)):
        return any(match_item(key, sub_value, item) for sub_value in value)
    else:
        return key not in item or str(item.get(key)).lower() == str(value).lower()


class ScreenShotsAPI(object):
    """
    Wrapper for BrowserStack Screenshots API.
    """
    root_url = 'https://www.browserstack.com/'
    default_browser = {
        'browser_version': '37.0',
        'os': 'Windows',
        'browser': 'firefox',
        'os_version': '8.1',
        'device': None
    }

    def __init__(self, user, key, default_browser=None, verbosity=0):
        self.auth = requests.auth.HTTPBasicAuth(user, key)
        self.default_browser = default_browser or self.default_browser
        self._cache = {}
        self.logger = get_logger(verbosity)
        self.logger.info('Username: %s; Password: %s;', user, key)
        self.logger.info('Default browser: %s;', self.default_browser)

    def execute(self, method, url, **kwargs):
        url = urljoin(self.root_url, url)
        self.logger.debug('Making "%s" request to "%s" with "%s"', method, url, str(kwargs))
        kwargs.setdefault('auth', self.auth)
        response = requests.request(method, url, **kwargs)
        self.logger.debug('Response: "%s"', response.content)
        response = response.json()
        if isinstance(response, dict):
            if response.get('message') == 'Parallel limit reached':
                raise ParallelLimitReached
            elif response.get('error') == 'Sign up or sign in':
                raise AuthError
        return response

    def browsers(self, browser=None, browser_version=None, device=None, os=None, os_version=None):
        """
        Returns list of available browsers & OS.
        """
        response = self.execute('GET', '/screenshots/browsers.json')
        for key, value in list(locals().items()):
            if key in ('self', 'response') or not value:
                continue
            response = [item for item in response if match_item(key, value, item)]
        return response

    def make(self, url, browsers=None, destination=None, timeout=DEFAULT_TIMEOUT, retries=DEFAULT_RETRIES, **kwargs):
        """
        Generates screenshots for given settings and saves it to specified destination.
        """
        response = self.generate(url, browsers, **kwargs)
        return self.download(response['job_id'], destination, timeout, retries)

    def generate(self, url, browsers=None, orientation=None, mac_res=None, win_res=None,
                             quality=None, local=None, wait_time=None, callback_url=None):
        """
        Generates screenshots for a URL.
        """
        if isinstance(browsers, dict):
            browsers = [browsers]
        if browsers is None:
            browsers = [self.default_browser]
        data = dict((key, value) for key, value in locals().items() if value is not None and key != 'self')
        return self.execute('POST', '/screenshots', json=data)

    def list(self, job_id):
        """
        Generate the list of screenshots and their states.
        """
        return self.execute('GET', '/screenshots/%s.json' % job_id)

    def download(self, job_id, destination=None, timeout=DEFAULT_TIMEOUT, retries=DEFAULT_RETRIES):
        """
        Downloads all screenshots for given job_id to `destination` folder.
        If `destination` is None, then screenshots will be saved in current directory.
        """
        self._retries_num = 0
        sleep(timeout)
        self.save_many(job_id, destination, timeout, retries)
        return self._cache

    def save_many(self, job_id, destination, timeout, retries):
        response = self.list(job_id)
        for screenshot in response['screenshots']:
            state = screenshot['state']
            if state == 'timed-out':
                continue
            if self._retries_num >= retries:
                self.logger.debug('Retry limit exceeded for %s' % job_id)
                break
            if state != 'done' or not screenshot['image_url']:
                self.logger.debug('Retrying download for %s' % job_id)
                self._retries_num += 1
                sleep(timeout)
                self.save_many(job_id, destination, timeout, retries)
                break
            self.save(screenshot['image_url'], destination)

    def save(self, image_url, destination=None):
        filename = image_url.split('/')[-1]
        if image_url in self._cache:
            return
        image_response = requests.get(image_url, stream=True)
        if destination:
            self.ensure_dir(destination)
            filename = os.path.join(destination, filename)
        self.logger.debug('Saving "%s" to "%s" ...', image_url, filename)
        self.save_file(filename, image_response)
        self._cache[image_url] = filename

    def ensure_dir(self, destination):
        """
        Checks, that `destination` exists.
        """
        try:
            os.makedirs(destination)
        except OSError:
            pass

    def save_file(self, filename, content):
        """
        Saves file on local filesystem.
        """
        with open(filename, 'wb') as f:
            for chunk in content.iter_content(chunk_size=1024):
                if chunk:
                    f.write(chunk)
